src/main/java/io/vertx/amqp/bridge/impl/AmqpConsumerImpl.java

```diff
@@ -0,0 +1,243 @@
+/*
+* Copyright 2016 the original author or authors.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+package io.vertx.amqp.bridge.impl;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import io.vertx.core.VertxException;
+import io.vertx.core.eventbus.Message;
+import io.vertx.core.eventbus.MessageConsumer;
+import io.vertx.core.eventbus.impl.BodyReadStream;
+import io.vertx.core.json.JsonObject;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import io.vertx.core.streams.ReadStream;
+import io.vertx.proton.ProtonConnection;
+import io.vertx.proton.ProtonReceiver;
+
+public class AmqpConsumerImpl implements MessageConsumer<JsonObject> {
```

Comment:
AmqpConsumerImpl can be accessed from non Vert.x thread so its members that are accessed from the Vert.x thread (proton callbacks) should be visible from both threads.

So we need to have some synchronization in the consumer for this. Have a look in the original implementation of `MessageConsumer` (HandlerRegistration).